staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,268 @@
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	golangproto "github.com/golang/protobuf/proto"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilresponsewriter "k8s.io/apiserver/pkg/util/responsewriter"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+const (
+	// groupVersionKindExtensionKey is the key used to lookup the
+	// GroupVersionKind value for an object definition from the
+	// definition's "extensions" map.
+	groupVersionKindExtensionKey = "x-kubernetes-group-version-kind"
+)
+
+// SchemaCacher fetches the openapi schema once and then caches it in memory
+type SchemaCacher struct {
+	rwMutex       sync.RWMutex
+	openAPISchema Resources
+	lastEtag      string
+
+	// These will never change once the cacher is set
+	downloader Downloader
+	parser     Parser
+}
+
+var _ Getter = &SchemaCacher{}
+
+// NewSchemaCacher creates a new SchemaCacher
+func NewSchemaCacher(handler http.Handler, path string) *SchemaCacher {
+	return &SchemaCacher{
+		downloader: &handlerDownloader{
+			handler: handler,
+			path:    path,
+		},
+		parser: &schemaParser{},
+	}
+}
+
+// Get implements Getter
+func (s *SchemaCacher) Get() (Resources, error) {
+	err := s.updateSchema()
+	if err != nil {
+		return nil, err
+	}
+
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.openAPISchema, nil
+}
+
+// updateSchema parses the OpenAPI schema if it has changed since the last
+// time updateSchema was called, and cache it.
+func (s *SchemaCacher) updateSchema() error {
+	lastEtag := s.getLastEtag()
+	schemaBytes, newEtag, httpStatus, err := s.downloader.DownloadSchema(lastEtag)
+	if err != nil {
+		return err
+	}
+
+	// The schema hasn't changed so there's nothing to update
+	if httpStatus == http.StatusNotModified {
+		return nil
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+
+	// Check etag again after taking the write lock in case lastEtag
+	// was updated while waiting for the write lock
+	if newEtag == s.lastEtag {
+		return nil
+	}
+
+	// Convert raw spec to our Resource type
+	schema, err := s.parser.Parse(schemaBytes)
+	if err != nil {
+		return err
+	}
+
+	s.openAPISchema = schema
+	s.lastEtag = newEtag
+	return nil
+}
+
+func (s *SchemaCacher) getLastEtag() string {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.lastEtag
+}
+
+// Downloader sepcifies where to download an OpenAPI schema from
+type Downloader interface {
+	// DownloadSchema attempts to get the OpenAPI schema if it has changed.
+	DownloadSchema(lastEtag string) (
+		schemaBytes []byte,
+		newEtag string,
+		httpStatus int,
+		err error)
+}
+
+type handlerDownloader struct {
+	handler http.Handler
+	path    string
+}
+
+var _ Downloader = &handlerDownloader{}
+
+// DownloadSchema implements Downloader
+func (d *handlerDownloader) DownloadSchema(lastEtag string) (schemaBytes []byte, newEtag string, httpStatus int, err error) {
```

Comment:
This downloader is less universal than the other components in this file. I'd consider putting it in its own file or package.